File: sdks/canary/python/scenarios/d_model_list_annotated.py
#!/usr/bin/env python3
# Copyright (C) 2026 Michael Kao
# SPDX-License-Identifier: AGPL-3.0-or-later
"""D-MODEL-LIST-ANNOTATED canary — Python SDK"""

from __future__ import annotations

import json
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from canary import jwt_login, Runner, Config, config_from_env, wait_active
from llmsafespaces import LLMSafeSpaces


def run(r: Runner, cfg: Config) -> None:
    c = LLMSafeSpaces(cfg.api_url, api_key=jwt_login(cfg), timeout=60.0)
    ws_id = None
    cred_id = None
    try:
        ok, ws = r.assert_no_error(
            lambda: c.workspaces.create(
                name="canary-py-model-list", runtime="base", storage_size="1Gi"
            ),
            "create: no error",
        )
        if not ok:
            return
        ws_id = ws.id

        phase = wait_active(c, ws_id)
        r.assert_(phase == "Active", "reach-active", f"got {phase!r}")
        if phase != "Active":
            return

        if cfg.llm_api_key:
            cred_value = json.dumps(
                {"kind": cfg.llm_provider, "slug": "canary-py-list-anno", "apiKey": cfg.llm_api_key}
            )
            ok_cred, cred = r.assert_no_error(
                lambda: c.secrets.create(
                    name="canary-py-model-cred", type="llm-provider", value=cred_value
                ),
                "create-cred: no error",
            )
            if ok_cred and cred:
                cred_id = cred.id
                r.assert_no_error(
                    lambda: c.workspaces.set_bindings(ws_id, [cred_id]),
                    "bind-cred: no error",
                )

        ok2, models_resp = r.assert_no_error(
            lambda: c.workspaces.get_models(ws_id), "get-models: no error"
        )
        if ok2 and models_resp is not None:
            r.assert_("models" in models_resp, "get-models: has models array")
            r.assert_("currentModel" in models_resp, "get-models: has currentModel")

            models = models_resp.get("models", [])
            r.assert_(len(models) > 0, "get-models: models non-empty")

            selected_count = 0
            for m in models:
                r.assert_("id" in m, f"model: has id")
                r.assert_("name" in m, f"model: has name")
                r.assert_("tier" in m, f"model: has tier")
                r.assert_("freeTier" in m, f"model: has freeTier")
                r.assert_("selected" in m, f"model: has selected")
                if m.get("selected"):
                    selected_count += 1

            r.assert_(
                selected_count >= 1,
                "get-models: at least one selected",
                f"found {selected_count} selected",
            )

        if cfg.llm_model:
            r.assert_no_error(
                lambda: c.workspaces.set_model(ws_id, cfg.llm_model),
                "set-model: no error",
            )
            ok3, models_after = r.assert_no_error(
                lambda: c.workspaces.get_models(ws_id), "get-models-after: no error"
            )
            if ok3 and models_after is not None:
                selected_models = [
                    m for m in models_after.get("models", []) if m.get("selected")
                ]
                r.assert_(
                    len(selected_models) >= 1,
                    "after-set: model is selected",
                    f"selected: {[m.get('id') for m in selected_models]}",
                )

    finally:
        if cred_id:
            try:
                c.secrets.delete(cred_id)
            except Exception:
                pass
        if ws_id:
            try:
                c.workspaces.delete(ws_id)
            except Exception:
                pass


if __name__ == "__main__":
    r = Runner("model-list-annotated")
    cfg = config_from_env()
    run(r, cfg)
    r.print()
    sys.exit(r.exit_code())

File: sdks/canary/typescript/scenarios/d-model-list-annotated.ts
// Copyright (C) 2026 Michael Kao
// SPDX-License-Identifier: AGPL-3.0-or-later
// D-MODEL-LIST-ANNOTATED canary — TypeScript SDK

import { LLMSafeSpaces } from '../../../typescript/src/index.js';
import { Runner, Config, configFromEnv, nodeFetch, waitActive } from '../canary.js';

async function run(r: Runner, cfg: Config): Promise<void> {
  if (!cfg.llmApiKey) { r.ok('model-list-annotated: skipped (no LLM API key)'); return; }
  const c = new LLMSafeSpaces({ baseUrl: cfg.apiUrl, apiKey: cfg.apiKey, timeout: 60000, fetch: nodeFetch as any });
  let wsId: string | null = null;
  try {
    const [ok, ws] = await r.assertNoError(
      () => c.workspaces.create({ name: 'canary-ts-model-list', runtime: 'base', storageSize: '1Gi' }),
      'create: no error');
    if (!ok || !ws) return;
    wsId = ws.id;

    const phase = await waitActive(c, wsId);
    r.assert(phase === 'Active', 'reach-active', `got "${phase}"`);
    if (phase !== 'Active') return;

    const [ok2, result] = await r.assertNoError(
      () => c.workspaces.getModels(wsId!), 'get-models: no error');
    if (ok2 && result) {
      r.assert(typeof result.currentModel === 'string', 'models: currentModel is string',
        typeof result.currentModel);
      r.assert(Array.isArray(result.models) && result.models.length > 0,
        'models: non-empty array', String(result.models?.length));

      if (Array.isArray(result.models)) {
        for (const m of result.models as any[]) {
          r.assert(typeof m.id === 'string', `model: has id`);
          r.assert(typeof m.name === 'string', `model: has name`);
          r.assert(typeof m.tier === 'string', `model: has tier`);
          r.assert(typeof m.selected === 'boolean', `model: has selected`);
        }

        const selected = (result.models as any[]).filter((m: any) => m.selected);
        r.assert(selected.length === 1, 'models: exactly one selected',
          `found ${selected.length}`);
        if (selected.length === 1) {
          r.assert(selected[0].id === result.currentModel,
            'models: selected id === currentModel',
            `${selected[0].id} vs ${result.currentModel}`);
        }
      }
    }

    await r.assertError(
      () => c.workspaces.getModels('00000000-0000-0000-0000-000000000000'),
      'get-models-nonexistent: error');

  } finally {
    if (wsId) { try { await c.workspaces.delete(wsId); } catch {} }
  }
}

async function main() {
  const r = new Runner('model-list-annotated');
  await run(r, configFromEnv());
  r.print();
  process.exit(r.exitCode());
}
main().catch(e => { console.error(e); process.exit(1); });
